kafka/consumer.go

```diff
@@ -395,6 +395,24 @@ func (c *Consumer) QueryWatermarkOffsets(topic string, partition int32, timeoutM
 	return queryWatermarkOffsets(c, topic, partition, timeoutMs)
 }
 
+// OffsetsForTimes looks up offsets by timestamp for the given partitions.
+//
+// The returned offset for each partition is the earliest offset whose
+// timestamp is greater than or equal to the given timestamp in the
+// corresponding partition.
+//
+// The timestamps to query are represented as `.Offset` in the `times`
+// argument and the looked up offsets are represented as `.Offset` in the returned
+// `offsets` list.
+//
+// The function will block for at most timeoutMs milliseconds.
+//
+// Duplicate Topic+Partitions are not supported.
+// Per-partition errors may be returned in the `.Error` field.
+func (c *Consumer) OffsetsForTimes(times []TopicPartition, timeoutMs int) (offsets []TopicPartition, err error) {
```

Comment:
Completely confusing having timestamps in a type called `TopicPartition`. 

The most natural data structure here is a map of TopicPartition : long, which is what the java client does. In dotnet I use an array also for consistency with the rest of the dotnet API, but on my todo list there is to change this where appropriate across the API.